plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/InlineMethodConverter.kt

```diff
@@ -5,26 +5,45 @@
 
 package org.jetbrains.kotlin.formver.conversion
 
+import org.jetbrains.kotlin.fir.expressions.FirLambdaArgumentExpression
 import org.jetbrains.kotlin.formver.embeddings.MethodSignatureEmbedding
 import org.jetbrains.kotlin.formver.embeddings.VariableEmbedding
 import org.jetbrains.kotlin.formver.viper.MangledName
 import org.jetbrains.kotlin.formver.viper.ast.Exp
 import org.jetbrains.kotlin.formver.viper.ast.Label
 
+sealed interface SubstitutionItem
+data class SubstitutionName(val name: MangledName) : SubstitutionItem
+data class SubstitutionLambda(val lambda: FirLambdaArgumentExpression) : SubstitutionItem
+
 class InlineMethodConverter(
     private val programCtx: ProgramConversionContext,
     override val signature: MethodSignatureEmbedding,
     override val returnVar: VariableEmbedding,
-    private val substitutionParams: Map<MangledName, MangledName>,
+    private val substitutionParams: Map<MangledName, SubstitutionItem>,
+    private val inlineNameAsDefault: Boolean = true // if false, names not in substitutionParams InlineName prefix is not added
 ) : MethodConversionContext, ProgramConversionContext by programCtx {
     override val returnLabel: Label = Label(InlineReturnLabelName(signature.name), listOf())
 
     // TODO: propagate the pre- and postconditions to here correctly.
     override val preconditions: List<Exp> = listOf()
     override val postconditions: List<Exp> = listOf()
-    override fun resolveName(name: MangledName): MangledName =
-        when {
+    override fun resolveName(name: MangledName): MangledName {
+        val sub = substitutionParams[name]
+        return when {
             name == ReturnVariableName -> returnVar.name
-            else -> substitutionParams[name] ?: InlineName(signature.name, name)
+            sub is SubstitutionName -> sub.name
+            else -> if (inlineNameAsDefault) InlineName(signature.name, name) else name
```

Comment:
Rather than distinguishing based on the type, I would recommend delegating this to a function of the interface.